src/riscv-integration.adoc

```diff
@@ -1078,27 +1079,75 @@ memory addresses must match for the address to be valid:
-not convert invalid addresses to other arbitrary invalid addresses in an
-unrestricted manner.
-The following procedure must be used instead when writing a capability A to
-these CSRs:
-
-. If A's address cannot be held then convert it to another address that the CSR can
-hold
-. If conversion _was_ required, then A's tag is cleared if A is
-sealed or if the new address is not representable -- this is equivalent to the
-semantics of <<SCADDR>>
+RISC-V permits that CSRs holding addresses, such as <<mtvec>> and <<mepc>>
+(see xref:CSR_exevectors[xrefstyle=short]) as well as <<pcc>>, need not
+hold all possible invalid addresses. Implementations may convert an invalid
+address into some other invalid address that the register is capable of holding.
+Therefore, implementations often support area and power optimizations by
+compressing invalid addresses in a lossy fashion.
+
+NOTE: Compressing invalid addresses allows implementations to reduce the number
+of flops required to hold some CSRs, such as <<mtvec>>. In CHERI, invalid
+addresses may also be used to reduce the number of bits to compare during a
+bounds check, for example, to 40 bits if using Sv39.
+
+However, the bounds encoding of capabilities in {cheri_base_ext_name} depends
+on the address value, so implementations must not convert invalid addresses to
+other arbitrary invalid address in an unrestricted manner. The remainder of
+this section describes how invalid address handling must be supported in
+{cheri_base_ext_name} when accessing CSRs, branching and jumping, and
+accessing memory.
+
+===== Accessing CSRs
+
+The following procedure must be used instead when executing instructions, such
+as <<CSRRW>>, that write a capability A to a CSR that cannot hold all invalid
+addresses:
+
+. If A's address is invalid and A does not have infinite bounds (see
+xref:section_cap_encoding[xrefstyle=short]), then A's tag is set to 0.
 . Write the final (potentially modified) version of capability A to the CSR e.g.
 <<mtvecc>>, <<mepcc>>, etc.
 
-This implies that sealed capabilities will always get their tags cleared when
-written to these CSRs unless the specification explicitly states that the CSR
-behaves otherwise (see <<mepcc>> and <<sepcc>>). Also note that <<pcc>> can
-be written with a <<JALR>> instruction in Capability Mode which automatically unseal the capability _before_
-the invalid address conversion above.
+===== Branches and Jumps
+
+The following procedure must be used when jumping or branching to a target
+capability A if the <<pcc>> cannot hold all invalid addresses. Note
+that the target capability A is the <<pcc>> when executing direct jumps and
+branches or the CHERI execution mode is Legacy (see
+xref:section_cheri_legacy_ext[xrefstyle=short]):
+
+. Calculate the effective target address T of the jump or branch as required by
+the instruction's behavior.
+. If T is invalid and A does not have infinite bounds (see
```

Comment:
Not sure how to add it here concisely, but this happens after tag/seal/perm checks but before bounds check.

Reply:
I added a list of priorities in the section where the codes are defined.